indexer/database/blocks.go

```diff
@@ -198,10 +200,37 @@ func (db *blocksDB) LatestL2BlockHeader() (*L2BlockHeader, error) {
 		if errors.Is(result.Error, gorm.ErrRecordNotFound) {
 			return nil, nil
 		}
-
 		return nil, result.Error
 	}
 
 	result.Logger.Info(context.Background(), "number ", l2Header.Number)
 	return &l2Header, nil
 }
+
+// Auxilliary Methods on both L1 & L2
+
+type Epoch struct {
+	L1BlockHeader L1BlockHeader `gorm:"embedded"`
+	L2BlockHeader L2BlockHeader `gorm:"embedded"`
+}
+
+// LatestEpoch return the latest epoch that has been seen on BOTH L1 & L2
+func (db *blocksDB) LatestEpoch() (*Epoch, error) {
+	// Since L1 blocks occur less frequently than L2, we do a INNER JOIN from L1 on
+	// L2 for a faster query. Per the protocol, the L2 block that starts a new epoch
+	// will have a matching timestamp with the L1 origin.
+	query := db.gorm.Table("l1_block_headers").Order("l1_block_headers.timestamp DESC")
+	query = query.Joins("INNER JOIN l2_block_headers ON l1_block_headers.timestamp = l2_block_headers.timestamp")
+	query = query.Select("*")
+
+	var epoch Epoch
+	result := query.Take(&epoch)
```

Comment:
Gorms API somewhat confuses me I get how it works but I don't get why it works this way (passing in a pointer)

Reply:
This isn't gorms api btw just passing a pointer into anything to fill up the pointer doesn't click with me sure it's a perfectly normal idiomatic thing though

Reply:
Lol it's been fun to learn Gorms API. Pretty intuitive but also a bit quirky in some situations

Also the reason the pointer is required is so that after the function returns, it's the same block of memory (struct backing the pointer) that's being updated as what the `epoch` variable is pointing to outside of the function call. If it didn't take a pointer, then the function would merely be updating a copy on the stack. The changes would not be seen in the intended `epoch` variable. Hope this makes some sense

Reply:
Also do be a bit more illustrative. It's not a requirement to take a reference of the first argument. It depends on the variable. For example this would work too

```go
epoch := new(Epoch)
result := query.Take(epoch)
```
This works because `new` allocates memory for the struct but returns a pointer. So `epoch` is already a point. Doing `Take(&epoch)` in this example would result in runtime error since &epoch is a pointer to a pointer

However this would **not** work
```go
var epoch *Epoch
result := query.Take(epoch)
```
This is because no memory has actually been allocated for an `Epoch` struct. In this state, epoch is simply a nil pointer typed to `*Epoch` but doesn't point anywhere. This would compile fun but result in a panic at runtime